// @ngx-pwa/offline/fesm5/offline.js
import { Inject, Injectable, InjectionToken, Optional, PLATFORM_ID, defineInjectable, inject } from '@angular/core';
import * as i0 from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { Router } from '@angular/router';
import * as i1 from '@angular/router';
import { empty, fromEvent, merge, of } from 'rxjs';
import { catchError, mapTo, startWith } from 'rxjs/operators';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var OFFLINE_ROUTE_OFFLINE = new InjectionToken('offline-config-route-offline', {
    providedIn: 'root',
    factory: function () { return '/offline'; }
});
var OFFLINE_ROUTE_UNAVAILABLE = new InjectionToken('offline-config-route-unavailable', {
    providedIn: 'root',
    factory: function () { return '/unavailable'; }
});
var OFFLINE_GUARDS_REDIRECT = new InjectionToken('offline-config-guards-redirect', {
    providedIn: 'root',
    factory: function () { return true; }
});
/**
 * @record
 */

/**
 * @param {?} config
 * @return {?}
 */
function offlineProviders(config) {
    return [
        config.routeOffline ? { provide: OFFLINE_ROUTE_OFFLINE, useValue: config.routeOffline } : [],
        config.routeUnavailable ? { provide: OFFLINE_ROUTE_UNAVAILABLE, useValue: config.routeUnavailable } : [],
        config.guardsRedirect === false ? { provide: OFFLINE_GUARDS_REDIRECT, useValue: false } : [],
    ];
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var Network = /** @class */ (function () {
    function Network(router, platformId, routeOffline, routeUnavailable) {
        this.router = router;
        this.platformId = platformId;
        this.routeOffline = routeOffline;
        this.routeUnavailable = routeUnavailable;
        /* Store instance in a static property to allow access to injected services in the RxJS static operator
             * Should be done only once */
        if (!Network.instance) {
            Network.instance = this;
        }
        this.initOnlineObservable();
    }
    Object.defineProperty(Network.prototype, "online", {
        /** Check if Internet connection is available */
        get: /**
         * Check if Internet connection is available
         * @return {?}
         */
        function () {
            return isPlatformBrowser(this.platformId) ? navigator.onLine : true;
        },
        enumerable: true,
        configurable: true
    });
    /** Do not use this method, use `catchOffline` function directly */
    /**
     * Do not use this method, use `catchOffline` function directly
     * @template T
     * @return {?}
     */
    Network.catchOffline = /**
     * Do not use this method, use `catchOffline` function directly
     * @template T
     * @return {?}
     */
    function () {
        return catchError(Network.catchCallback);
    };
    /**
     * @template T
     * @param {?} error
     * @param {?} caught
     * @return {?}
     */
    Network.catchCallback = /**
     * @template T
     * @param {?} error
     * @param {?} caught
     * @return {?}
     */
    function (error, caught) {
        if (!Network.instance) {
            console.log("You need to import OfflineModule in your AppModule\n      AND to inject the Network service at least once, for example in your AppComponent constructor.");
            throw error;
        }
        else if (!Network.instance.router) {
            console.log("You need to import RouterModule.forRoot([]) in your application.");
            throw error;
        }
        else {
            if (!Network.instance.online) {
                Network.instance.router.navigate([Network.instance.routeOffline]);
                return empty();
            }
            else if (error.status && (error.status >= 500 && error.status < 600)) {
                Network.instance.router.navigate([Network.instance.routeUnavailable]);
                return empty();
            }
            else {
                throw error;
            }
        }
    };
    /**
     * @return {?}
     */
    Network.prototype.initOnlineObservable = /**
     * @return {?}
     */
    function () {
        this.onlineChanges = !isPlatformBrowser(this.platformId) ? of(true) : merge(fromEvent(window, 'online').pipe(mapTo(true)), fromEvent(window, 'offline').pipe(mapTo(false)))
            .pipe(startWith(this.online));
    };
    Network.instance = null;
    Network.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] },
    ];
    /** @nocollapse */
    Network.ctorParameters = function () { return [
        { type: Router, decorators: [{ type: Optional },] },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] },] },
        { type: undefined, decorators: [{ type: Inject, args: [OFFLINE_ROUTE_OFFLINE,] },] },
        { type: undefined, decorators: [{ type: Inject, args: [OFFLINE_ROUTE_UNAVAILABLE,] },] },
    ]; };
    /** @nocollapse */ Network.ngInjectableDef = defineInjectable({ factory: function Network_Factory() { return new Network(inject(Router, 8), inject(PLATFORM_ID), inject(OFFLINE_ROUTE_OFFLINE), inject(OFFLINE_ROUTE_UNAVAILABLE)); }, token: Network, providedIn: "root" });
    return Network;
}());
/**
 * Catch offline errors (no Internet connection) and server errors (HTTP status 5xx)
 * and redirect to /offline or /unavailable page (routes can be changed in the OfflineModule)
 */
var catchOffline = Network.catchOffline;

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Check if Internet connection is available to allow a navigation to a route or not.
 * By default, if Internet connection is not available, the user will be redirected to the /offline page.
 * This behavior and the URL of redirection can be configured in the OfflineModule.
 */
var OnlineGuard = /** @class */ (function () {
    function OnlineGuard(router, network, routeOffline, guardsRedirect) {
        this.router = router;
        this.network = network;
        this.routeOffline = routeOffline;
        this.guardsRedirect = guardsRedirect;
    }
    /**
     * @return {?}
     */
    OnlineGuard.prototype.canActivate = /**
     * @return {?}
     */
    function () {
        return this.guard();
    };
    /**
     * @return {?}
     */
    OnlineGuard.prototype.canActivateChild = /**
     * @return {?}
     */
    function () {
        return this.guard();
    };
    /**
     * @return {?}
     */
    OnlineGuard.prototype.canLoad = /**
     * @return {?}
     */
    function () {
        return this.guard();
    };
    /**
     * @return {?}
     */
    OnlineGuard.prototype.guard = /**
     * @return {?}
     */
    function () {
        if (!this.network.online) {
            if (this.guardsRedirect) {
                this.router.navigate([this.routeOffline]);
            }
            return false;
        }
        return true;
    };
    OnlineGuard.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] },
    ];
    /** @nocollapse */
    OnlineGuard.ctorParameters = function () { return [
        { type: Router, },
        { type: Network, },
        { type: undefined, decorators: [{ type: Inject, args: [OFFLINE_ROUTE_OFFLINE,] },] },
        { type: undefined, decorators: [{ type: Inject, args: [OFFLINE_GUARDS_REDIRECT,] },] },
    ]; };
    /** @nocollapse */ OnlineGuard.ngInjectableDef = defineInjectable({ factory: function OnlineGuard_Factory() { return new OnlineGuard(inject(Router), inject(Network), inject(OFFLINE_ROUTE_OFFLINE), inject(OFFLINE_GUARDS_REDIRECT)); }, token: OnlineGuard, providedIn: "root" });
    return OnlineGuard;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { offlineProviders, OFFLINE_GUARDS_REDIRECT, OFFLINE_ROUTE_OFFLINE, OFFLINE_ROUTE_UNAVAILABLE, Network, catchOffline, OnlineGuard };
